pkg/apis/crds/karpenter.k8s.aws_ec2nodeclasses.yaml

```diff
@@ -143,18 +155,41 @@ spec:
                             and sc1: 125-16,384 \n * standard: 1-1,024"
                           pattern: ^(\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))(([KMGTPE]i)|[numkMGTPE]|([eE](\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))))?$
                           x-kubernetes-int-or-string: true
+                          x-kubernetes-validations:
+                          - rule: self.endsWith('Gi') || self.endsWith('Ti') || self.endsWith('G')
+                              || self.endsWith('T')
+                          - rule: 'self.endsWith(''Gi'') || self.endsWith(''G'') ?
+                              double(self.find(''[0-9]*'')) >= 1.0 : true'
+                          - rule: 'self.endsWith(''Gi'') || self.endsWith(''G'') ?
+                              double(self.find(''[0-9]*'')) <= 6400.0 : true'
+                          - rule: 'self.endsWith(''Ti'') || self.endsWith(''T'') ?
+                              double(self.find(''[0-9]*'')) >= 0.001 : true'
+                          - rule: 'self.endsWith(''Ti'') || self.endsWith(''T'') ?
+                              double(self.find(''[0-9]*'')) <= 64.0 : true'
                         volumeType:
                           description: VolumeType of the block device. For more information,
                             see Amazon EBS volume types (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumeTypes.html)
                             in the Amazon Elastic Compute Cloud User Guide.
+                          enum:
+                          - standard
+                          - io1
+                          - io2
+                          - gp2
+                          - sc1
+                          - st1
+                          - gp3
                           type: string
                       type: object
+                      x-kubernetes-validations:
+                      - message: snapshotID or volumeSize must be defined
+                        rule: has(self.snapshotID) || has(self.volumeSize)
                     rootVolume:
                       description: RootVolume is a flag indicating if this device
                         is mounted as kubelet root dir. You can configure at most
                         one root volume in BlockDeviceMappings.
                       type: boolean
                   type: object
+                maxItems: 15
```

Comment:
Just curious. Is there a maximum number of block device mappings that are enforced by EC2? If so, should we enforce the limit that they impose? I think 15 may be a reasonable amount but curious if we just adopt what they say.

Reply:
I was not able to find an exact number. From what I can find the maximum number of volumes that can be attached for an instance can vary on the type.